client/verta/tests/unit_tests/pipeline/test_pipeline_graph.py

```diff
@@ -0,0 +1,162 @@
+
+import verta
+from tests.unit_tests.strategies import pipeline_definition
+from verta.pipeline import PipelineGraph
+
+
+def test_set_steps(make_mock_pipeline_step, make_mock_registered_model) -> None:
+    """Test that the steps of a PipelineGraph can be set."""
+    mocked_rm = make_mock_registered_model(id=123, name="test_rmv")
+    with patch.object(
+        verta.pipeline.PipelineStep, "_get_registered_model", return_value=mocked_rm
+    ):
+        step_1 = make_mock_pipeline_step()
+        step_2 = make_mock_pipeline_step()
+        graph = PipelineGraph(steps=set())
+    graph.set_steps({step_1, step_2})
+    assert set(graph.steps) == {step_1, step_2}
+    graph.set_steps(set())
+    assert not graph.steps
+
+
+@given(
+    pipeline_definition=pipeline_definition(),
+    registered_model_id=st.integers(min_value=1, max_value=2**63),
+    # max value limit avoids protobuf "Value out of range" error
+    model_version_name=st.text(min_size=1),
+    model_name=st.text(min_size=1),
+)
+@settings(
+    suppress_health_check=[HealthCheck.function_scoped_fixture],
+    deadline=None,
+)
+def test_from_definition(
+    pipeline_definition,
+    registered_model_id,
+    model_version_name,
+    model_name,
+    mock_conn,
+    mock_config,
+    mocked_responses,
+) -> None:
+    """Test that a PipelineGraph object can be constructed from a pipeline
+    definition.
+
+    The model version is fetched for each step, so a response
+    is mocked for each.  In depth testing of each step is handled in
+    test_pipeline_step.test_steps_from_pipeline_spec.
+    """
+    for step in pipeline_definition["steps"]:
+        mocked_responses.get(
+            f"https://test_socket/api/v1/registry/model_versions/{step['model_version_id']}",
+            json={
+                "model_version": {
+                    "id": step["model_version_id"],
+                    "registered_model_id": registered_model_id,
+                    "version": model_version_name,
+                }
+            },
+            status=200,
+        )
+        mocked_responses.get(
+            f"https://test_socket/api/v1/registry/registered_models/{registered_model_id}",
+            json={
+                "registered_model": {
+                    "id": registered_model_id,
+                    "name": model_name,
+                }
+            },
+            status=200,
+        )
+    graph = PipelineGraph._from_definition(
+        pipeline_definition=pipeline_definition, conn=mock_conn, conf=mock_config
+    )
+    # the object produced is a PipelineGraph
+    assert isinstance(graph, PipelineGraph)
+    # we have the same number of steps as in the pipeline definition
+    assert len(graph.steps) == len(pipeline_definition["steps"])
+    # sort each group of steps for comparison
+    pipeline_steps_sorted = sorted(
+        pipeline_definition["steps"], key=lambda x: x["name"]
+    )
+    graph_steps_sorted = sorted(list(graph.steps), key=lambda x: x.name)
```

Comment:
We actually don't need to cast a `set` to a `list` before passing it to `sorted()`

```suggestion
    graph_steps_sorted = sorted(graph.steps, key=lambda x: x.name)
```